shark_turbine/kernel/wave/wave_sim.py

```diff
@@ -184,19 +221,71 @@ def decorator(func: Callable[..., Any]) -> Any:
 
 
 def _read_proxy(
-    memory: "Memory", elements_per_thread: Optional[IndexExpr] = None
+    memory: "Memory",
+    elements_per_thread: Optional[IndexExpr] = None,
+    mapping: Optional[IndexMapping] = None,
 ) -> "Register":
-    return memory.clone()
+    if mapping:
+        input_sym_shape = _get_symbolic_shape(memory)
+        inp_mapping = mapping.map_input_indices(input_sym_shape)
+        res_mapping = mapping.map_output_indices()
+        iters = mapping.iters
+
+        def mapping_func(ind_mapping, indices):
+            subs = [(ind, val) for ind, val in zip(iters, indices)]
+            return _to_indices(ind.subs(subs) for ind in ind_mapping)
+
+        iter_shape = mapping.iter_shape
+        res_shape = mapping.out_shape
+
+        res = torch.zeros(res_shape)
+        _set_symbolic_shape(res, mapping.output_shape)
+        for index in np.ndindex(*iter_shape):
+            inp_mapped = mapping_func(inp_mapping, index)
+            res_mapped = mapping_func(res_mapping, index)
+            res[res_mapped] = memory[inp_mapped]
+
+    else:
+        res = memory.clone()
+        _set_symbolic_shape(res, _get_symbolic_shape(memory))
+
+    return res
 
 
 def _write_proxy(
-    src: "Register", dst: "Memory", elements_per_thread: Optional[IndexExpr] = None
+    src: "Register",
+    dst: "Memory",
+    elements_per_thread: Optional[IndexExpr] = None,
+    mapping: Optional[IndexMapping] = None,
 ) -> None:
+    if mapping:
+        input_sym_shape = _get_symbolic_shape(src)
+        inp_mapping = mapping.map_input_indices(input_sym_shape)
+        res_mapping = mapping.map_output_indices()
+        iters = mapping.iters
+
+        def mapping_func(ind_mapping, indices):
+            subs = [(ind, val) for ind, val in zip(iters, indices)]
+            return _to_indices(ind.subs(subs) for ind in ind_mapping)
+
+        iter_shape = mapping.iter_shape
+
+        for index in np.ndindex(*iter_shape):
+            inp_mapped = mapping_func(inp_mapping, index)
+            res_mapped = mapping_func(res_mapping, index)
+            dst[res_mapped] = src[inp_mapped]
+
+        return
+
     dst[:] = src
 
 
 def _mma_proxy(a: "Register", b: "Register", acc: "Register") -> "Register":
-    return torch.matmul(a, b.T) + acc
+    a_shape = _get_symbolic_shape(a)

```

Comment:
Do we need to set a_shape, b_shape and res shape? Doesn't seem like they are used here?

Reply:
Th reason I need those `_get/_set_symbolic_shape` calls, because memory/register arrays are represented by `torch.tensor` and getting shape from it will return actual shape, but I also want to know original symbolic shapes that were associated with those arrays. For now, I'm just maintaining a separate map `id(object) -> symbolic_shape` for simplicity. Another approach will be to create a wrapper on top of `torch.tensor` and keep symbolic shape there.

Reply:
Sounds good, thanks!